### ISSUE TITLE
[Discuss] ObserverComponent render props return type

### ISSUE THREAD

**Issue by inomdzhon** [mobxjs/mobx-react-lite#305]
Can we use `React.ReactNode` type for return type of `ObserverComponent`?

https://github.com/mobxjs/mobx-react-lite/blob/4157a0c44607d73b9a6a99975ec4527e08095465/src/ObserverComponent.ts#L4

**Comment by mweststrate** [comment]
@inomdzhon would you mind including a motivation?

**Comment by inomdzhon** [comment]
@mweststrate I apologize, sure.

In some situation we need return empty value (`null`/`undefined`, etc) by condition, example
```
class State {
  @observable isLoading = true;
  @observable count = 0;
}

<Observer>
   {() => state.isLoading ? null : <div>{state.count}</div>}
</Observer>
```
or maybe we want return `string`/`number`/`array` (is valid react element)
```
<Observer>
   {() => state.isLoading ? null : state.count}
</Observer>
```

Type of `React.ReactNode` is `ReactChild | ReactFragment | ReactPortal | boolean | null | undefined`.

**Comment by mweststrate** [comment]
Check, double checking, React.ReactElement doesn't include `null`? In that
case feel free to open a PR

**Comment by FredyC** [comment]
`ReactNode` is kinda wrong as it includes `undefined` which isn't supported return value from rendering. This is a baggage of the past that cannot be simply changed because of class components shenanigans.

**Comment by inomdzhon** [comment]
@mweststrate no, it doesn't include `null`

---

@FredyC oh, I don't know about 😱  I use `React.ReactNode` in many place... What about if we exclude `undefined`?

Use TS utility `Exclude<React.ReactNode, undefined>` or set own typing `ReactChild | ReactFragment | ReactPortal | boolean | null`

**Comment by mweststrate** [comment]
In that case `React.ReactElement | null` sounds like a simpler solution?

**Comment by inomdzhon** [comment]
Hm, `ReactChild | ReactFragment | ReactPortal | boolean | null` will not be superfluous.

This issue arose after we updated `mobx-react` from 4 to 5 version. In version 4 return type was `React.ReactNode`.

![image](https://user-images.githubusercontent.com/5850354/91299791-df000b00-e7aa-11ea-9994-bb91059263e2.png)

**Comment by mweststrate** [comment]
Ideally keep the change as small as possible. Not sure what boolean is
doing there, pretty sure that Fragment is also covered by Element? If you
are looking for null, which makes sense, just add that one? Not sure what
changed between 4 and 5, my memory doesn't go that far ;-)

**Comment by inomdzhon** [comment]
Sorry, I edited message – from 5 to 6 version)

> Fragment is also covered by Element?

It is does. `ReactPortal` covered too. Exclude `ReactFragment` and `ReactPortal`.

> Not sure what boolean is doing there

Agree, exclude `boolean`.

---

Then `ReactChild | null` 👌  ?

**Comment by mweststrate** [comment]
Sounds ok. Feel free to open a PR :)

**Comment by FredyC** [comment]
> Not sure what boolean is doing there

It's mostly for the short-circuit reasons when you use `&&` operator which yields `false`. React can handle that properly and it's comparable to returning `null`. Not sure about strings though, probably something of the past too.

**Comment by mweststrate** [comment]
rarely know, but React components are allowed to directly return a string,
in case you just want to render some text

**Comment by inomdzhon** [comment]
Create PR https://github.com/mobxjs/mobx-react-lite/pull/307

**Comment by mweststrate** [comment]
Released as 2.2.0
